sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -0,0 +1,446 @@
+   *    transformed during physical planning. For example:
+   *     Agg           HashAgg          HashAgg
+   *      |               |                |
+   *    child    =>     Xchg      =>     Stage1
+   *                      |
+   *                   HashAgg
+   *                      |
+   *                    child
+   *    The updated plan node will be:
+   *    LogicalQueryStage(HashAgg - Stage1)
+   */
+  private def updateLogicalPlan(
+      logicalPlan: LogicalPlan,
+      physicalPlan: SparkPlan,
+      newStages: Seq[(Exchange, QueryStageExec)]): LogicalPlan = {
+    var currentLogicalPlan = logicalPlan
+    newStages.foreach { case (exhange: Exchange, stage: QueryStageExec) =>
+      // Get the corresponding logical node for `exchange`. If `exchange` has been transformed from
+      // a `Repartition`, it should have `logicalLink` available by itself; otherwise traverse down
+      // to find the first node that is not generated by `EnsureRequirements`.
+      val logicalNodeOpt = exhange.logicalLink.orElse(exhange.collectFirst {
+        case p if p.logicalLink.isDefined => p.logicalLink.get
+      })
+      assert(logicalNodeOpt.isDefined)
+      val logicalNode = logicalNodeOpt.get
+      val physicalNode = physicalPlan.collectFirst {
+        case p if p.eq(stage) || p.logicalLink.exists(logicalNode.eq) => p
+      }
+      assert(physicalNode.isDefined)
+      // Replace the corresponding logical node with LogicalQueryStage
+      val newLogicalNode = LogicalQueryStage(logicalNode, physicalNode.get)
+      var cnt = 0
+      val newLogicalPlan = currentLogicalPlan.transformDown {
+        case p if p.eq(logicalNode) =>
+          cnt += 1
+          newLogicalNode
+      }
+      if (cnt > 1) {
+        throw AmbiguousLogicalMappingException(logicalNode)
+      }
+      assert(cnt == 1,
+        s"logicalNode: $logicalNode; " +
+          s"logicalPlan: $currentLogicalPlan " +
+          s"physicalPlan: $physicalPlan" +
+          s"stage: $stage")
+      currentLogicalPlan = newLogicalPlan
+    }
+    currentLogicalPlan
+  }
+
+  /**
+   * Re-optimize and run physical planning on the current logical plan based on the latest stats.
+   */
+  private def reOptimize(logicalPlan: LogicalPlan): (SparkPlan, LogicalPlan) = {
+    logicalPlan.invalidateStatsCache()
+    val optimized = optimizer.execute(logicalPlan)
+    SparkSession.setActiveSession(session)
+    val sparkPlan = session.sessionState.planner.plan(ReturnAnswer(optimized)).next()
+    val newPlan = applyPhysicalRules(sparkPlan, queryStagePreparationRules)
+    (newPlan, optimized)
+  }
+
+  /**
+   * Notify the listeners of the physical plan change.
+   */
+  private def onUpdatePlan(): Unit = {
+    executionId.foreach { id =>
+      val exec = SQLExecution.getQueryExecution(id)
+      if (exec != null) {
+        session.sparkContext.listenerBus.post(SparkListenerSQLAdaptiveExecutionUpdate(
+          id,
+          exec.toString,
```

Comment:
Do we need to print the LogicalPlans?

Reply:
Not sure I understand it. The physical plan has a link to the logical plan, is that sufficient?

Reply:
nvm. This is to show all the plans, including parsed, analyzed, optimized and physical.